### DISCUSSION TITLE
move vs copy when emplacing a context variable

### DISCUSSION THREAD

**Discussion by dpunsetAGS** [skypjack/entt#866] [Q&A]
Hi,

I have a really weird bug that I'd like to share, in case it rings any bell to someone. I'll start by saying that I'm trying to use entt in an unreal project, so I have targets "Development Editor" and "Development".

I just started using context variables as a way to store singleton state data. Everything was working ok in "Development Editor", but then switching to "Development" it crashes.

I've spent some time debugging, and I noticed that the reason for the crash is that when emplacing a context variable it reallocates a vector, and for some reason it's deciding to copy a previously emplaced context variable instead of moving it. The problem is, this context variable is not movable at all because it has a unique_ptr inside.

To me what's really surprising is having a different behavior in those two targets, also considering that this is all done at compile time.

Looking at the callstack when things go well and things go wrong, the difference is that in the good one std::_Uninitialized_move will be called, and on the bad one std::_Uninitialized_copy. Down the line, in the bad case it will reach basic_any::basic_vtable with the copy operation which will do nothing, because it's compiled out with the if constexpr (so this part detects is_copy_constructible_v fine)

These are the callstacks for the good and bad case, same code but changing from “Development Editor” to “Development”

GOOD
```
MyProject.dll!entt::basic_any<0,8>::basic_vtable<MyFirstEmplacedContextVariable>(const entt::basic_any<0,8>::operation op, const entt::basic_any<0,8> & value, const void * other) Line 56           C++
MyProject.dll!entt::basic_any<0,8>::basic_any<0,8>(entt::basic_any<0,8> && other) Line 195       C++
MyProject.dll!entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>::dense_map_node<unsigned int const ,entt::basic_any<0,8>>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> && __that)   C++
MyProject.dll!std::_Uninitialized_backout_al<std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>::_Emplace_back<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> && <_Vals_0>) Line 1632             C++
MyProject.dll!std::_Uninitialized_move<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> *,std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * const _First, entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * const _Last, entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * _Dest, std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>> & _Al) Line 1733         C++
MyProject.dll!std::vector<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>,std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>::_Emplace_reallocate<unsigned __int64 &,std::piecewise_construct_t const &,std::tuple<unsigned int const &>,std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&>>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * const _Whereptr, unsigned __int64 & <_Val_0>, const std::piecewise_construct_t & <_Val_1>, std::tuple<unsigned int const &> && <_Val_2>, std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&> && <_Val_3>) Line 794          C++
MyProject.dll!std::vector<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>,std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>::emplace_back<unsigned __int64 &,std::piecewise_construct_t const &,std::tuple<unsigned int const &>,std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&>>(unsigned __int64 & <_Val_0>, const std::piecewise_construct_t & <_Val_1>, std::tuple<unsigned int const &> && <_Val_2>, std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&> && <_Val_3>) Line 747          C++
MyProject.dll!entt::dense_map<unsigned int,entt::basic_any<0,8>,entt::identity,std::equal_to<unsigned int>,std::allocator<std::pair<unsigned int const ,entt::basic_any<0,8>>>>::get_or_emplace<std::piecewise_construct_t const &,std::tuple<unsigned int const &>,std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&>>(const unsigned int & key, const std::piecewise_construct_t & <args_0>, std::tuple<unsigned int const &> && <args_1>, std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&> && <args_2>) Line 295         C++
MyProject.dll!entt::dense_map<unsigned int,entt::basic_any<0,8>,entt::identity,std::equal_to<unsigned int>,std::allocator<std::pair<unsigned int const ,entt::basic_any<0,8>>>>::try_emplace<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t>(const unsigned int & key, const std::in_place_type_t<MySecondEmplacedContextVariable> & <args_0>, void * && <args_1>) Line 633  C++
MyProject.dll!entt::internal::registry_context::emplace_hint<MySecondEmplacedContextVariable,std::nullptr_t>(const unsigned int id, void * && <args_0>) Line 164      C++
```

BAD
```
MyProject.exe!entt::basic_any<0,8>::basic_vtable<MyFirstEmplacedContextVariable>(const entt::basic_any<0,8>::operation op, const entt::basic_any<0,8> & value, const void * other) Line 56           C++
MyProject.exe!entt::basic_any<0,8>::basic_any<0,8>(const entt::basic_any<0,8> & other) Line 181            C++
MyProject.exe!entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>::dense_map_node<unsigned int const ,entt::basic_any<0,8>>(const entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> & __that)     C++
MyProject.exe!std::_Uninitialized_backout_al<std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>::_Emplace_back<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> &>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> & <_Vals_0>) Line 1632             C++
MyProject.exe!std::_Uninitialized_copy<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> *,std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * const _First, entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * const _Last, entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * _Dest, std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>> & _Al) Line 1672         C++
MyProject.exe!std::vector<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>,std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>::_Emplace_reallocate<unsigned __int64 &,std::piecewise_construct_t const &,std::tuple<unsigned int const &>,std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&>>(entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>> * const _Whereptr, unsigned __int64 & <_Val_0>, const std::piecewise_construct_t & <_Val_1>, std::tuple<unsigned int const &> && <_Val_2>, std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&> && <_Val_3>) Line 794          C++
MyProject.exe!std::vector<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>,std::allocator<entt::internal::dense_map_node<unsigned int const ,entt::basic_any<0,8>>>>::emplace_back<unsigned __int64 &,std::piecewise_construct_t const &,std::tuple<unsigned int const &>,std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&>>(unsigned __int64 & <_Val_0>, const std::piecewise_construct_t & <_Val_1>, std::tuple<unsigned int const &> && <_Val_2>, std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&> && <_Val_3>) Line 747          C++
MyProject.exe!entt::dense_map<unsigned int,entt::basic_any<0,8>,entt::identity,std::equal_to<unsigned int>,std::allocator<std::pair<unsigned int const ,entt::basic_any<0,8>>>>::get_or_emplace<std::piecewise_construct_t const &,std::tuple<unsigned int const &>,std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&>>(const unsigned int & key, const std::piecewise_construct_t & <args_0>, std::tuple<unsigned int const &> && <args_1>, std::tuple<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t &&> && <args_2>) Line 295         C++MyProject.exe!entt::dense_map<unsigned int,entt::basic_any<0,8>,entt::identity,std::equal_to<unsigned int>,std::allocator<std::pair<unsigned int const ,entt::basic_any<0,8>>>>::try_emplace<std::in_place_type_t<MySecondEmplacedContextVariable> const &,std::nullptr_t>(const unsigned int & key, const std::in_place_type_t<MySecondEmplacedContextVariable> & <args_0>, void * && <args_1>) Line 633  C++
MyProject.exe!entt::internal::registry_context::emplace<MySecondEmplacedContextVariable,std::nullptr_t>(void * && <args_0>) Line 169                C++
```

I’ve tried reproducing the same thing in a small standalone project in case I could post it but I was not able to reproduce it. I guess the issue must be something very specific to the way the projects are compiled.

One thing I don’t understand is how would get_or_emplace on the dense_map know if the type is copyable or not, because all it knows is entt::basic_any<0,8>. It doesn’t explain why in one target it chooses move and in the other copy, but could it be that dense_map just doesn’t support non-copyable types?

Any ideas?

Thank you!

**Comment by dpunsetAGS** [comment]
A friend has found what seems to cause this change in behavior. One difference between "Development Editor" and "Development" is that the first has exceptions enabled and the other does not. When exceptions are enabled, the emplace will move, and when they are disabled the emplace will copy, causing the bug.

In order to fix it, we did this:

```
#if defined(__cpp_exceptions) && !defined(ENTT_NOEXCEPTION)
#    define ENTT_THROW throw
#    define ENTT_TRY try
#    define ENTT_CATCH catch(...)
#else
#    define ENTT_THROW
#    define ENTT_TRY if(true)
#    define ENTT_CATCH if(false)
#endif

#define ENTT_NOEXCEPT noexcept
```

With this it seems to move always.
I was able to reproduce the bug with my simple test case out of the unreal context.

Any thoughts on this?

**Comment by skypjack** [comment]
> could it be that dense_map just doesn’t support non-copyable types?

It does actually. The dense map is a kind of glorified and customized sparse set with a fixed size sparse array. It swaps the elements internally to keep the packed array really packed all the time. So, both movable and copyable objects are fine. Of course, movable types perform way better, as you can guess.
As for the `entt::any`, it's movable and copyable but it really depends on the underlying object. If you try to copy an `entt::any` and the contained type isn't copyable, it returns an empty (and thus invalid) object.

Let's look at your callstack. The _issue_ arises when the internal vector runs out of memory and decides to reallocate due to an `emplace_back`.
From what I see (you are with MSVC, right?), this is the required condition to provide a strong exception guarantee:
```cpp
if constexpr (is_nothrow_move_constructible_v<_Ty> || !is_copy_constructible_v<_Ty>) {
    // move
} else {
    // copy
}
```
Therefore, everything seems to boil down to `dense_map_node` and `basic_any`. If we look at the latter, its move constructor is marked as `ENTT_NOEXCEPT` and you said that brute-force setting `ENTT_NOEXCEPT` to `noexcept` (no, this isn't an option for a default configuration in `EnTT` 😅) solves the problem.
So, maybe `is_nothrow_move_constructible_v` always looks at the `noexcept`-ness of a type even when exceptions are disabled? If this is the case, it's definitely annoying.

**Comment by skypjack** [comment]
Yeah, confirmed, `is_nothrow_move_constructible_v` works the same regardless of whether exceptions are enabled or not. This is true with all major compilers.
So, to sum up, what happens is that `dense_map_node` has a default copy and move constructors. It means that their `noexcept`-ness derive from that of the data members. `next` is an `std::size_t` and therefore it's irrelevant here. `element` is an `std::any` for the registry context. In this case, the move constructor is `ENTT_NOEXCEPT`, that is, _conditionally_ `noexcept`. Therefore, when you (well, UE) disable exceptions, the move constructor of `entt::any` is no longer `noexcept` because of this:
```cpp
#if defined(__cpp_exceptions) && !defined(ENTT_NOEXCEPTION)
```
As a consequence, the move constructor of `dense_map_node` is no longer `noexcept` and `is_nothrow_move_constructible_v` is false. Since `dense_map_node` is (required to be) copy constructible, the whole guard here is false:
```cpp
if constexpr (is_nothrow_move_constructible_v<_Ty> || !is_copy_constructible_v<_Ty>) {
```
The fallback is then to copy the underlying value. Very tricky.

**Comment by skypjack** [comment]
Probably, the whole `ENTT_NOEXCEPT` thing is _wrong_. It was introduced because of #61 but I don't know if it still makes sense at all.
From what you write, it doesn't seem so actually. You redefined `ENTT_NOEXCEPT` to always be `noexcept`, even when UE compiles with disabled exceptions.

**Comment by skypjack** [comment]
In the meantime, I've created an issue [here](https://github.com/skypjack/entt/issues/867#issue-1205379288) out of this discussion. To be discussed further though.
